### ISSUE TITLE
Malformed client certificate subject error unclear

### ISSUE THREAD

**Issue by Billy99** [rustls/rustls#1292]
With a rust based client/server (https://github.com/bpfd-dev/bpfd), the rust server code generates a set of certificates using the openssl::x509 crate. A GO client we use is able to use the certificate with the rust server, but the rust client fails with the same certificate. Using OpenSSL to created the certificates, both the GO client and the rust client succeed.

```
$ RUST_LOG=debug bpfctl list
:
[2023-05-01T18:27:53Z DEBUG rustls::anchors] add_parsable_certificates processed 1 valid and 0 invalid certs
[2023-05-01T18:27:53Z DEBUG hyper::client::connect::http] connecting to [::1]:50051
[2023-05-01T18:27:53Z DEBUG hyper::client::connect::http] connected to [::1]:50051
[2023-05-01T18:27:53Z DEBUG rustls::client::hs] No cached session for DnsName("localhost")
[2023-05-01T18:27:53Z DEBUG rustls::client::hs] Not resuming any session
[2023-05-01T18:27:53Z DEBUG rustls::client::hs] Using ciphersuite TLS13_AES_256_GCM_SHA384
[2023-05-01T18:27:53Z DEBUG rustls::client::tls13] Not resuming
[2023-05-01T18:27:53Z DEBUG rustls::client::tls13] TLS1.3 encrypted extensions: [Protocols([ProtocolName(6832)]), ServerNameAck]
[2023-05-01T18:27:53Z DEBUG rustls::client::hs] ALPN protocol is Some(b"h2")
[2023-05-01T18:27:53Z DEBUG rustls::client::tls13] Got CertificateRequest CertificateRequestPayloadTLS13 { context: , extensions: [SignatureAlgorithms([ECDSA_NISTP384_SHA384, ECDSA_NISTP256_SHA256, ED25519, RSA_PSS_SHA512, RSA_PSS_SHA384, RSA_PSS_SHA256, RSA_PKCS1_SHA512, RSA_PKCS1_SHA384, RSA_PKCS1_SHA256]), AuthorityNames([DistinguishedName(30123110300e06035504030c07627066642d6361)])] }
[2023-05-01T18:27:53Z DEBUG rustls::client::common] Attempting client auth
Error: transport error

Caused by:
    0: error trying to connect: invalid peer certificate: BadEncoding
    1: invalid peer certificate: BadEncoding
```

I haven't been able to pin point if it is a rustls issue or not, but wanted to raise the alarm in case a release was pending.

Using `sudo openssl x509 -in client.pem -text -noout` to dump both the OpenSSL and rust generated certs (ca.pem and client.pem), they appear to equivalent (execpt the hex encoded portions).

**Comment by cpu** [comment]
Hi @Billy99,

Would you be able to provide an example certificate that fails with a Rustls client? It would save some reproduction effort.

**Comment by Billy99** [comment]
Yes, was working on it. The raw files or the `sudo openssl x509 -in client.pem -text -noout` output?

**Comment by cpu** [comment]
Raw files please :-) PEM or DER would be fine.

**Comment by Billy99** [comment]
<details><summary>Rust generated bpfd-client.pem:</summary>

```
-----BEGIN CERTIFICATE-----
MIIFJzCCAw+gAwIBAgIUcfroorCTk8PR9zBG4p9H20xosqowDQYJKoZIhvcNAQEL
BQAwEjEQMA4GA1UEAwwHYnBmZC1jYTAeFw0yMzA1MDExOTMzNDdaFw0yMzA1MzEx
OTMzNDdaMBYxFDASBgNVBAMMC2JwZmQtY2xpZW50MIICIjANBgkqhkiG9w0BAQEF
AAOCAg8AMIICCgKCAgEAn3r5K6Txo84bWkCykNu/w5l4TGceV/8QRtLM0H/yV1xS
AtdjkjKYr59tR/F7u9Hipt0gQRoCy/Hn8sku+VUASApSTNKRN/2lyUnz/5Msa5w1
19KwWNkmWxbeXUaESEpU9aUoK21/2cPAXZr5VSLTj8UkvW7syilPyvkPEuEQ94nv
n9tagCn2nKsxbsLVCfJFH2VajUjd0ycpw8cO8aHHov5ZoSaGDArUv1a3zuH4Gfj3
MdluQVUUWA6JAGPtFsYh0MMPb/vIO8ASAUwcbXiIe1wZ207NHtMLM7QRQpyC2c6/
Biv/byDgtSCkCaacl+gJmoBj4mXGy8C0QPKcgYqgwJppTkKElN2Bu94ZZyTbqggu
zWZZzfGY5L0Y0lreoON3zptFoQ1u8TteAQoRhf42nhUCiW2qMESe1La0+Sm/dfkR
oXZdcYc1De/CzeepHoeQnltu1Uk3YP36mDZR1yrpLM31kcZejz9wH73fTFnVNNIN
5sLoUxyRAo64j26q9NDnfbNk64rT3Ak9uTASFVZb2KX3pWhM8KjhE9ITi+S/Becc
FTXub37FfRcdAWfi+Lh3lIy/1nNCNg9vwp6YqyrSpeWBrTaFvWm7TGWR5yzcMBsA
H+1Cn2jIh4MPi5gxO/wMIARx2LVxYuPVQBbStcTAN1fw8SVGsfsBu26RpUfp+C0C
AwEAAaNxMG8wLQYDVR0RBCYwJIIXbG9jYWxob3N0LCBJUDoxMjcuMC4wLjGCCWxv
Y2FsaG9zdDAdBgNVHQ4EFgQUbChr8pjYQCjElqye3UIuYdsQzucwHwYDVR0jBBgw
FoAU4JsuTN0Gyz3rQ8ImTHj80GVwXOswDQYJKoZIhvcNAQELBQADggIBABglgOwU
kU+j/QtpuJ6Ho2EXfRsYTm5l5ueiWURckC2BNuTrm+eF4dI89Xb3mqXK1+IX6HFd
Jm7gW3KqjSZDRkgKajFA2GEvQncvJfCSQ8CFl6omJ+icgIpoTjc3bj2YEl1LXu//
4DqZ59OoWeZ9YJ0497KwkozhOJW7e3XZKByveKvn4PJ78oMGxafWp9OWZ3e++Knf
1ibPFVYM3MSj6cVthaC0s4+61FUicvX4OKVsX8E1ehnA5Jg0GcS3bvt51Fs/py0l
A67ECegIkECiYn8HKgh8oHfCArbPkU/R4tf54dwEwSpJFkXfFJx35f+Hu1QkQdE2
grw3u9w8cpIQi8Pe9bNZ7s2CsrdLJUEth6+zOfIHO8TNF24P+4q9uUMd0AxtidFN
dRwvc4JZKRGjSmF7ZrhKu59J/N/z/qp4lMYH4giJDX1XdtrxjK1uyIOaO9xqgvoF
Dl6EwW4jp/taxDqwHGN9eg+465Z20jBCIychtjrLZ2MXXR/rY6lVIfkALYEiz4sJ
jcUp9hl5exGx4zRhgR3LvU5pYOBzDyavUtm0SlCDbw1yeVQqjbZp/nVNzdoJHT7q
pX97sCREm6MN+nAGcEcTPhh2v9PYARaMKWJmB/tH9HSW6ZaoV6WgAWdWCoAGNmXB
p2IAns/a30dIlJfGFr9cswXiYEmV7wXldkat
-----END CERTIFICATE-----
```
</detaiis>

**Comment by Billy99** [comment]
<details><summary>OpenSSL generated bpfd-client.pem:</summary>

```
-----BEGIN CERTIFICATE-----
MIIFFDCCAvygAwIBAgIUPHMoSztJxlMDyJdIvKHrZssm/D4wDQYJKoZIhvcNAQEL
BQAwEjEQMA4GA1UEAwwHYnBmZC1jYTAeFw0yMzA0MjgyMTEzMDhaFw0yMzA1Mjgy
MTEzMDhaMBYxFDASBgNVBAMMC2JwZmQtY2xpZW50MIICIjANBgkqhkiG9w0BAQEF
AAOCAg8AMIICCgKCAgEAtJ2lKV68cG0gH042BiaJZACDb/QyEMoOEjcUeqMrfrWF
M6XnK5vkD9VmwPjjkIz3aRQ2XR41RvDaU4wNpy43HHn1HwzNbMzZJrp3yZgmTlxx
IkKzhz5Fbn6Nkk14jUcaN94TTKS4FXRfxpkU1rKAy4MT539LD06GvY8Ke4dtUsO/
XgX1lvM6lDUzE50spPeleW+mlgeMjJOcHrkhEJ5w0jKEebKvu2YpexRuh8X1tA/d
Fmvzc1e9ip3ATVgwUkWJw4bJw9LJqU2m9zLEtfjk6R0HRySMP3JlaEvc3ObiCZ38
6DbQ4fi56HqdMTNOCkMUWRfWb+IVcJ/TJWXy1HJJgTMLRGUbCIkkHhdeGz26IW7A
TwgQtpVp6+Df0Sotzszu/a3qkRBSgDXDYIvDlx1DYBHRj/6t9UGakA88Gsy6/2sZ
Z/hop7OG04pd6B9b9CETahEU6Z9rGSrWoR2FhdlvHe6HalUXMcm+1yS58UKMMeW3
522zHHHhMVICyKnGBVKO3c7ZAi73meDbWI0b0h/NBrkMtByeJZxhl7MuCzn5mrqa
sSfdZQiAlyG/6RsaKmPleq54ut37nsQA6N8xIC6osWfWS306OAtmAe530ZRyQjtk
6fctF1gyIWS+p8JmA1Q2zc/UknwDu9dgJDMBnkrOkyBQr3OwJAAinZDmLWEdwn0C
AwEAAaNeMFwwGgYDVR0RBBMwEYIJbG9jYWxob3N0hwR/AAABMB0GA1UdDgQWBBR3
pLiVUBZtV3ZtRH6u6WhzVye+MjAfBgNVHSMEGDAWgBS46JwEwHJM35xKhSEVMoYZ
xLhzNjANBgkqhkiG9w0BAQsFAAOCAgEALFq3aYIY5aiwh+bmR7Kl6C9/CXiZ5MbX
3dtILzUPKnenIqD7G700Ido/QHhNoilB+d1nM0mrZ7g8Y/ZrMY5Du1mtV3my843/
Q00NATEsQHjVKuyuAOIF8PNKkHGSRR5qXTlS2h/+Lrksj54nASznWVirbeMLCeL4
wIrZ6aO1LwOyT3yS9eeLSVSmqDvzWzsBs8bL9Q/6JVTagUNBntekAsEkYWV8q2RQ
GYxA8VYpGTwBGuScXDdrcUvf5Z+OXu6AYLRYN7Pu1JRICOPXr2ufoONdNl4K3iNO
RgEm4+bb6KOtTcaJjs4b0D0gXXYFbVOj0KuMu+b629AfAiZAHO4nnAq9CHq/gBZy
cdMWkZnvQy1ebuRDbgA6IzWGgd3wf63qoDSApOFRnA9qvbALxW/JuuXetSinEHcm
cGJyqZqsNXbs1VqU2piGOhhTRUIvnukX+Td+DX2FSuViptTk/VxeWzAbkINQ/aTS
yOUod2QX0M0HwgGBH+4EJtoTjAKnqdWNUJU2R3fabExa/bkL+tgSt5FvnuHqsACH
NlTJrtIknm96O8DqTOj26oxwTWRgcWQOdrbWZYmBEhS3JxWejKg0TOFYydNn/SPZ
gVVdnJ9YIEVpKa+GJZSBP+QOuFLz6QZ5+0oDtzpXk9YzieFlpvAMloe3/R5XpiKx
qqjAM9YkR3o=
-----END CERTIFICATE-----
```
</detaiis>

**Comment by Billy99** [comment]
Let me know what else I can provide. I have the ca.pem files as well if you need them.

**Comment by cpu** [comment]
> Rust generated bpfd-client.pem:

Off-hand I notice the Rust generated certificate has duplicate subject alternate name (SAN) extensions for `localhost` while the OpenSSL generated one does not:
> DNS:localhost, IP:127.0.0.1, DNS:localhost

I'm not confident that's the problem but it might be prudent to remove the duplication and see if it makes any difference.

**Comment by Billy99** [comment]
I have. I actually added that recently because the GO Client (with the recent bump of tonic/OpenSSL/rustls) started failing with:
```
$ ./go-xdp-counter -iface vethdd340d3
2023/05/01 14:28:01 Using Input: Interface=vethdd340d3 Priority=50 Source=0xc0003706f0
2023/05/01 14:28:01 Reading /etc/bpfd/gocounter.toml ...
2023/05/01 14:28:01 Read /etc/bpfd/gocounter.toml failed: err open /etc/bpfd/gocounter.toml: no such file or directory
2023/05/01 14:28:01 rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate is valid for localhost, IP:127.0.0.1, not localhost"
```

**Comment by Billy99** [comment]
But you bring up a good point with the duplicate subject alternate name (SAN) extensions, because the OpenSSL generated certificates doesn't have the duplicate, and it works. So that is another issue, but don't know if that is an issue with rustls or tonic. (Sorry, I'm not an expert in this TLS, just hacking my way through it).

**Comment by cpu** [comment]
Aha! I think I see the problem.

> failed to verify certificate: x509: certificate is valid for localhost, IP:127.0.0.1, not localhost

The `openssl x509` text output is deceiving us and this error message has the problem identified.

I believe one of the SANs is literally `"localhost, IP:127.0.0.1"`. It's not a DNS type SAN and a separate IP type SAN. It's _one_ DNS type SAN that has an invalid value with a comma, the label "IP:" and an IP address. There is no IP type SAN at all in the Rust generated certificate.

**Comment by cpu** [comment]
Here's a quick Go snippet demonstrating what I mean: https://go.dev/play/p/aczsbgiFUGs

**Comment by Billy99** [comment]
So the error is in how the certificates were generated. Any thoughts on why this was working and stopped working with the recent upgrades (not important, just curious)?

**Comment by cpu** [comment]
> So the error is in how the certificates were generated

That's my intuition right now but I have only looked at the certificates and not tried to replicate end-to-end using Rustls as a client.

>  Any thoughts on why this was working and stopped working with the recent upgrades (not important, just curious)?

That's a good question. I have to drop offline in a moment but I'd be interested in trying a more realistic test replication with my own certificate + private key to see if I can figure that out. I can look at trying that ~tomorrow if nobody scoops me :-)

If this is the correct root cause it would be nice if we could perhaps surface a better error message to save others some debugging effort.

**Comment by Billy99** [comment]
I have to step away as well, but I did rebuild the code real fast, and:
```
    let subject_alt_name = SubjectAlternativeName::new()
-        .dns("localhost, IP:127.0.0.1")
+        .dns("localhost")
        .build(&cert_builder.x509v3_context(Some(&ca_cert_x590), None))?;
```
Works for GO and Rust clients. The original used to work.

**Comment by cpu** [comment]
> Any thoughts on why this was working and stopped working with the recent upgrades (not important, just curious)?

@Billy99 I can't resist a good mystery and I found a few minutes to test this. I stepped back ~3 Rustls releases and all of them correctly rejected the malformed client certificate. The exact error has changed with 0.21.x but the behaviour to reject the malformed certificate hasn't. That made me think the source of the change was likely on the certificate generation side and `rust-openssl`. It must have been the case that at some point your code was generating a valid certificate.

I was able to trace the change in behaviour to rust-openssl [`openssl-v0.10.48`](https://github.com/sfackler/rust-openssl/releases/tag/openssl-v0.10.48). Specifically it was https://github.com/sfackler/rust-openssl/pull/1854/commits/482575bff434f58b80ffea34a9610d0ff265ac1f  Prior to that injection vulnerability fix using `dns("localhost, IP:127.0.0.1")` would create two SAN extensions: a DNS one for `localhost` and an IP address one for `127.0.0.1`. In v0.10.48+ you get one malformed DNS SAN instead. I suspect the fact that it worked correctly in <0.10.48 was an accident, and relates to the danger of the vuln that was fixed.

You should be able to get a correctly generated certificate using rust-openssl v0.10.48+ by specifying one `dns()` extension and one `ip()` extension like:
```rust
let subject_alt_name = SubjectAlternativeName::new()
        .dns("localhost")
        .ip("127.0.0.1")
        ....
```

I think that resolves the mystery fully, but still leaves the question of whether we can improve the error message Rustls generates. We should give that more thought.

**Comment by Billy99** [comment]
I'm fine with closing this issue unless you want to leave it open to give more thought on improving the error message.

**Comment by cpu** [comment]
> I'm chasing this down in the webpki crate (https://github.com/rustls/webpki/issues/59) but I believe we'll also have to make a change on the Rustls side once that lands to surface a more specific error.

Circling back on this now that we're using the updated webpki.

We did resolve [webpki#59](https://github.com/rustls/webpki/issues/59) with more specific errors for misencoded DNS subject names (and name constraints), but there's no change needed on the Rustls side because in https://github.com/rustls/webpki/pull/69 we made the change to ignore these two specific errors when validating an end entity certificate for a subject name, continuing iteration to look for a valid matching identifier. 

The net result is that instead of getting a `rustls::error::BadEncoding` error for the class of problem reported here webpki will yield `webpki::error::CertNotValidForName` after running out of possible well-formed subject names to compare against. The webpki error for this condition will be mapped to `rustls::error::CertificateError::NotValidForName`. That seems pretty sensible to me. 

Closing this out since I don't think there's any work left to do that wouldn't be undoing the work from [webpki#69](https://github.com/rustls/webpki/pull/69)
